### PR TITLE
feat: notification support stack

### PR DESCRIPTION
**Pull request by MadCcc**
- feat: notification stack
- feat: notification support stack

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
RFC: https://github.com/ant-design/ant-design/discussions/44359

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      Notification support `stack` props, more than 3 notifications will be stacked by default now.     |
| 🇨🇳 Chinese |     Notification 组件支持 `stack` 配置，默认超过三个以上的提示会堆叠起来。      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 3269707</samp>

This pull request adds the stacked notifications feature to the `notification` component. It allows the user to configure the stacking behavior and appearance of the notifications using the `stack` property. It also refactors the notification styles and animations to support different placements and alignments. It updates the `rc-notification` dependency to the latest version. It modifies the files `components/notification/interface.ts`, `components/notification/style/index.ts`, `components/notification/useNotification.tsx`, `components/notification/style/placement.ts`, `components/notification/style/stack.ts`, and `package.json`.

### <samp>🤖 Generated by Copilot at 3269707</samp>

*  Replace `NotificationPlacement` type with constant array and derived type ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-c80bd08ac7a1f6a8ae2046fe84cc562807b2d11b1a63316286546b888a86ed08L7-R15))
*  Import `genStackStyle` function from `stack.ts` to generate CSS styles for stacked notifications ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965R7), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-afcb3078a4ca46e4c23d5a6292b9d6bddc597e289711b77fc319bad238d5b85fR1-R164))
*  Add `notificationStackLayer` property to `NotificationToken` interface to represent maximum number of stacked notifications ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965R33))
*  Rename and move `notificationFadeIn` keyframes to `placement.ts` and use different keyframes for different placements ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965L61-R64), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-da4d2980e968bf7ec54db139303120e6105d73f3d214f60e9f5feec3b9c60dbdL7-R102), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-afcb3078a4ca46e4c23d5a6292b9d6bddc597e289711b77fc319bad238d5b85fR1-R164))
*  Wrap notice element with new element and use `noticeCls` variable for class name ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965L101-R95), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965L262-R236), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-da4d2980e968bf7ec54db139303120e6105d73f3d214f60e9f5feec3b9c60dbdL7-R102))
*  Move `genNotificationPlacementStyle` function to `placement.ts` and remove `animationName` property from fade-in animation rules ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965L204-R195), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965L236-R221), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-da4d2980e968bf7ec54db139303120e6105d73f3d214f60e9f5feec3b9c60dbdL7-R102))
*  Simplify fade-out animation rules and rename keyframes to `fadeOut` ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965L236-R221))
*  Remove comment about index.less variables ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965L285))
*  Add `notificationStackLayer` property to `notificationToken` object and return array of style objects from `genNotificationStyle` function ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-1b72cd8f976d21d83517b2c963a0b9a9bb6670e5d004ed4a7b0dd2edad3fc965L295-R275))
*  Import `useToken` hook from `theme/internal` module and declare `token` variable ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-fa8b6f08e906ed18759c056ccb358fb0b780b4ffbb0f00ed368b45c3b8efe5dfR21), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-fa8b6f08e906ed18759c056ccb358fb0b780b4ffbb0f00ed368b45c3b8efe5dfL65-R69))
*  Pass `stack` property to `Holder` component and `rc-notification` component and use `token.margin` value for gap ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-fa8b6f08e906ed18759c056ccb358fb0b780b4ffbb0f00ed368b45c3b8efe5dfL65-R69), [link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-fa8b6f08e906ed18759c056ccb358fb0b780b4ffbb0f00ed368b45c3b8efe5dfR95-R102))
*  Update `rc-notification` dependency version to `~5.2.0-0` in `package.json` ([link](https://github.com/ant-design/ant-design/pull/44618/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L139-R139))

### CODE REVIEW

**Review by li-jia-nan — APPROVED** [review]

**Comment by afc163** [comment]
怎么大了这么多？

**Comment by yoyo837** [comment]
> 怎么大了这么多？

https://github.com/react-component/notification/compare/v5.1.1...v5.2.0-2 和 此PR都加了不少东西

**Comment by afc163** [comment]
> https://github.com/react-component/notification/compare/v5.1.1...v5.2.0-2 和 此PR都加了不少东西

看 diff 感觉不应该有这么大。

**Review by kiner-tang — APPROVED** [review]

**Comment by kiner-tang** [comment]
> > [react-component/notification@v5.1.1...v5.2.0-2](https://github.com/react-component/notification/compare/v5.1.1...v5.2.0-2) 和 此PR都加了不少东西
> 
> 看 diff 感觉不应该有这么大。

有没有可能其他的一些底层依赖更新了 patch 版本导致的

**Comment by MadCcc** [comment]
@afc163 
bundle 问题看这个 https://github.com/react-component/notification/pull/315

File: components/notification/index.zh-CN.md [hunk 1]
```diff
@@ -75,6 +76,7 @@ config 参数如下：
 | getContainer | 配置渲染节点的输出位置 | () => HTMLNode | () => document.body |  |
 | placement | 弹出位置，可选 `top` `topLeft` `topRight` `bottom` `bottomLeft` `bottomRight` | string | `topRight` |  |
 | rtl | 是否开启 RTL 模式 | boolean | false |  |
+| stack | 堆叠模式，超过阈值时会将所有消息收起 | boolean \| `{ threshold: number }` | `{ threshold: 3 }` | 5.10.0 |
```
Comment (li-jia-nan) [inline components/notification/index.zh-CN.md:79]: 这个要 5.10.0 才发吗？

Comment (kiner-tang) [reply components/notification/index.zh-CN.md:79]: 5.9.0 还没开车的话，是不是可以挤上去

Comment (yoyo837) [reply components/notification/index.zh-CN.md:79]: 下一个minor是5.9.0噢

Comment (MadCcc) [reply components/notification/index.zh-CN.md:79]: 预计是 5.10，能上车再改

File: package.json [hunk 2]
```diff
@@ -136,7 +136,7 @@
     "rc-mentions": "~2.7.0",
     "rc-menu": "~9.12.0",
     "rc-motion": "^2.7.3",
-    "rc-notification": "~5.2.0-2",
+    "rc-notification": "~5.2.0-3",
```
Comment (kiner-tang) [inline package.json:139]: 如果没其他问题的话，是不是发一个正式版本

Comment (yoyo837) [reply package.json:139]: 大佬说,都不着急敢5.9.0了, 所以可以多测测, 正式版也就不着急了.

Comment (MadCcc) [reply package.json:139]: 嗯，准备 merge 的时候再发，防止有其他问题需要 patch

File: package.json [hunk 3]
```diff
@@ -137,7 +137,7 @@
     "rc-mentions": "~2.7.0",
     "rc-menu": "~9.12.0",
     "rc-motion": "^2.9.0",
-    "rc-notification": "~5.1.1",
+    "rc-notification": "~5.2.0-4",
```
Comment (afc163) [inline package.json:140]: 搞个正式版

**Pull request merged**